core/src/network/handler/query.rs

```diff
@@ -5,31 +5,85 @@ use crate::{
     instance::dispatch_action,
     network::query::{
         GetLinksNetworkQuery, GetLinksNetworkResult, NetworkQuery, NetworkQueryResult,
+        GetLinkData
     },
     nucleus,
 };
-use holochain_core_types::{crud_status::CrudStatus, entry::EntryWithMetaAndHeader};
+use holochain_core_types::{crud_status::CrudStatus, entry::EntryWithMetaAndHeader,error::HolochainError,eav::Attribute};
 use holochain_json_api::json::JsonString;
 use holochain_net::connection::json_protocol::{QueryEntryData, QueryEntryResultData};
 use holochain_persistence_api::cas::content::Address;
-use std::{collections::BTreeSet, convert::TryInto, sync::Arc};
+use std::{convert::TryInto, sync::Arc};
 
 fn get_links(
     context: &Arc<Context>,
     base: Address,
     link_type: String,
     tag: String,
     crud_status: Option<CrudStatus>,
-) -> Vec<(Address, CrudStatus)> {
-    context
+    headers : bool
+) -> Result<Vec<GetLinkData>,HolochainError> {
+    //get links
+    let dht_store = context
         .state()
         .unwrap()
-        .dht()
+        .dht();
+    
+    let (get_link ,error) : (Vec<_>,Vec<_>) = dht_store
         .get_links(base, link_type, tag, crud_status)
-        .unwrap_or(BTreeSet::new())
+        .unwrap_or_default()
         .into_iter()
-        .map(|eav_crud| (eav_crud.0.value(), eav_crud.1))
-        .collect::<Vec<_>>()
+        .map(|eav_crud| (eav_crud.0.value(), eav_crud.1,eav_crud.0.attribute()))
+        //get tag
+        .map(|eav_crud|{
+            let tag = match eav_crud.2
+            {
+                Attribute::LinkTag(_,tag) => Ok(tag),
+                Attribute::RemovedLink(_,tag) => Ok(tag),
+                _ =>  Err(HolochainError::ErrorGeneric("Could not get tag".to_string()))
+            }.expect("INVALID ATTRIBUTE ON EAV GET, SOMETHING VERY WRONG IN EAV QUERY");
+            (eav_crud.0,eav_crud.1,tag)
+        })
+        //get targets from dht
+        .map(|eav_crud|{
+            dht_store
+            .get_link_targets(eav_crud.0.clone())
```

Comment:
I'm not sure if it is safe to be getting the targets from the store like this. With full sync there is a good chance that the target are held by the same node as the base but this might not be the case with sharding right? Maybe there is something I am not understanding

Reply:
I think with this situation because of how every node that stores the base also stores the meta and target(check dht_reduers_inner line 59), this should be fine. If no base is found in the local dht it won't even reach this far

Reply:
@willemolding You were right with this on the aspect that getting the targets like this would work because of the link_caching but getting the entries would be the problem. Just cause one has the link doesn't mean they have the linkadd entry(is a part I overlooked) switched this around so take a look when you can!